### PR TITLE
Add LAN/WAN modes and enhance installer mode detection with tests

### PR DESCRIPTION
**Pull request by jumpsmm7**
## Pull Request Summary

This PR introduces explicit **WAN and LAN installation modes** and carries those modes throughout installer configuration, runtime behavior, service lifecycle management, and migration paths.

### Key changes

- **Adds router-mode detection.** Router mode (`sw_mode=1`) maps to WAN, while repeater, access-point, and media-bridge modes (`sw_mode=2`, `3`, or `4`) map to LAN only when a usable LAN IPv4 address is available. Unknown or invalid NVRAM values remain unconfirmed instead of silently forcing a mode. [installerL1978-L2010](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/67013742c797b3b2da5fc5c8d64ffa7fb9e6536e/installer#L1978-L2010)
- **Persists mode-specific defaults.** WAN installs retain IPSet and managed dnsmasq behavior, while LAN installs disable IPSet, default dnsmasq management to automatic, and optionally use the LAN gateway as the reverse-DNS upstream. [installerL1797-L1843](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/67013742c797b3b2da5fc5c8d64ffa7fb9e6536e/installer#L1797-L1843)
- **Aligns runtime defaults with the detected mode.** New installations persist both `ADGUARD_INSTALL_MODE` and the matching LAN/WAN netcheck mode; upgrades preserve legacy defaults for compatibility. [installerL1846-L1875](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/67013742c797b3b2da5fc5c8d64ffa7fb9e6536e/installer#L1846-L1875)
- **Adds runtime mode helpers and guards.** Runtime scripts can determine whether they are operating in LAN mode, avoid IPSet there, and manage dnsmasq only when appropriate. [AdGuardHome.shL186-L221](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/67013742c797b3b2da5fc5c8d64ffa7fb9e6536e/AdGuardHome.sh#L186-L221)
- **Changes LAN networking behavior.** LAN mode skips public WAN connectivity probes but continues checking the local AdGuardHome DNS listener. [README.mdL285-L299](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/67013742c797b3b2da5fc5c8d64ffa7fb9e6536e/README.md#L285-L299)
- **Scopes LAN listeners more carefully.** LAN mode binds loopback, the primary LAN bridge address, eligible stable IPv6, and qualifying RFC 1918 secondary bridge addresses while excluding temporary, invalid, public, and other unsafe addresses. [README.mdL301-L324](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/67013742c797b3b2da5fc5c8d64ffa7fb9e6536e/README.md#L301-L324)
- **Separates listener and firewall policy.** LAN mode does not install firewall/IPTABLES rules, making listener selection independent from traffic-access policy. [README.mdL323-L327](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/67013742c797b3b2da5fc5c8d64ffa7fb9e6536e/README.md#L323-L327)
- **Expands regression coverage.** Tests cover WAN/router, repeater, AP, media-bridge, unknown, missing, wildcard, loopback, multicast, and invalid-address detection cases. [tests/installer-install-mode-detection.shL490-L535](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/67013742c797b3b2da5fc5c8d64ffa7fb9e6536e/tests/installer-install-mode-detection.sh#L490-L535)
- **Verifies LAN/WAN IPSet separation.** Regression coverage confirms LAN mode skips IPSet locks, rewrites, and restarts while preserving the existing WAN path. [tests/ipset-lan-mode.shL115-L142](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/67013742c797b3b2da5fc5c8d64ffa7fb9e6536e/tests/ipset-lan-mode.sh#L115-L142)

The PR also includes broader lifecycle, rollback, portability, security-scanning, checksum, release-validation, and automated review infrastructure hardening reflected in its larger final commit.

**Checks performed**

- ✅ `git status --short --branch`
- ✅ `git diff --stat d1bcb9dcd3ba5e67a30328c1dcaeb7bad0d33c6a..67013742c797b3b2da5fc5c8d64ffa7fb9e6536e`
- ✅ `git diff --name-status d1bcb9dcd3ba5e67a30328c1dcaeb7bad0d33c6a..67013742c797b3b2da5fc5c8d64ffa7fb9e6536e`
- ✅ `git log --oneline --decorate d1bcb9dcd3ba5e67a30328c1dcaeb7bad0d33c6a..67013742c797b3b2da5fc5c8d64ffa7fb9e6536e`

No files were changed for this summary-only request, so no commit or follow-up pull request was created.

 [View task →](https://chatgpt.com/s/cd_6a8959d285248191bfbfb0e2d67177f5)

### CODE REVIEW

**Comment by jumpsmm7** [comment]
@codex resolve all feedback

File: tests/fixtures/portability/fail/echo-e.sh [hunk 1]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+echo -e 'bad\noutput'
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/echo-e.sh:2]: @qodo-ai "that is what this test is trying to show."

File: tests/fixtures/portability/fail/pipefail.sh [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+set -o pipefail
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/pipefail.sh:2]: @qodo-ai "that is what this test is trying to show."

File: tests/fixtures/portability/fail/select.sh [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+select value in one two; do break; done
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/select.sh:2]: @qodo-ai "that is what this test is trying to show."

File: tests/fixtures/portability/fail/mapfile.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+mapfile values <input
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/mapfile.sh:2]: @qodo-ai "that is what this test is trying to show."

File: tests/fixtures/portability/fail/process-substitution.sh [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+cat <(printf '%s\n' bad)
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/process-substitution.sh:2]: @qodo-ai "that is what this test is trying to show."

File: tests/fixtures/portability/fail/read-array.sh [hunk 6]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+read -r -a values
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/read-array.sh:2]: @qodo-ai "that is what this test is trying to show."

File: AdGuardHome.sh [hunk 7]
```diff
@@ -1156,52 +1783,398 @@ proc_optimizations() {
...
+			proc_lock_claim_acquire "${self_start}" || exit 1
+			if mkdir "${PROC_LOCK_DIR}" 2>/dev/null; then
+				trap 'proc_lock_mkdir_cleanup; exit 1' HUP INT QUIT ABRT TERM TSTP
+				proc_lock_claim_matches "$$" "${self_start}" || {
+					proc_lock_mkdir_cleanup
+					exit 1
+				}
+				printf '%s %s\n' "$$" "${self_start}" >"${PROC_LOCK_DIR}/pid" || {
+					proc_lock_mkdir_cleanup
+					exit 1
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh:2006]: /q you can dismiss this as an issue because all downloads have to pass both sha256, and md5sum verification significantly reducing the risk of a MITM attack. It is documented that --insecure is only invoked on legacy routers where a security certificate may no longer be valid.

File: tests/fixtures/portability/fail/mapfile.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/mapfile.sh:2]: @qodo-ai "this is a test to demonstrate that mapfile is invalid"

File: tests/fixtures/portability/fail/pipefail.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/pipefail.sh:2]: @qodo-ai "this is a test to demonstrate that pipefail is invalid"

File: tests/fixtures/portability/fail/select.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/select.sh:2]: @qodo-ai dismiss this.

File: tests/fixtures/portability/fail/pipefail.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/pipefail.sh:2]: @qodo-ai dismiss this issue

File: tests/fixtures/portability/fail/mapfile.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/fixtures/portability/fail/mapfile.sh:2]: @qodo-ai dismiss this issue

**Comment by jumpsmm7** [comment]
@qodo-ai "dismiss function keyword in fixture"

**Comment by jumpsmm7** [comment]
@qodo-ai "dismiss issues # 12 - 15"

**Comment by jumpsmm7** [comment]
@qodo-ai "dismiss issues # 4"

**Comment by jumpsmm7** [comment]
@codex review the complete PR for final merge readiness, including correctness, security, BusyBox ash compatibility, failure safety, interruption recovery, rollback behavior, permissions, checksums, CI coverage, missing (or broken/failing/non-up-to-date) regression tests, and edge cases

**Comment by jumpsmm7** [comment]
@codex review the complete PR for final merge readiness, including correctness, security, BusyBox ash compatibility, failure safety, interruption recovery, rollback behavior, permissions, checksums, CI coverage, missing (or broken/failing/non-up-to-date) regression tests, and edge cases

**Comment by jumpsmm7** [comment]
@codex
I'll retrieve the logs for this failing job to identify the issue and provide a solution.Now let me retrieve the workflow file and the code quality test file to understand what's being tested:## Analysis

The job failed at the final step with exit code 1. Looking at the logs, the failure occurs after completing all regression tests and ShellCheck analysis, ending at the "shfmt mksh formatting check" step.

The critical error is on line 2026-08-22T23:05:18.4042616Z:
```
##[error]Process completed with exit code 1.
```

This happens after the "shfmt mksh formatting check" completes with "OK: shfmt mksh formatting check", but before the workflow step "Explain formatting remediation" can run.

## Root Cause

The issue is in the [code-quality.sh](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/e01455be6b2da14953feb31299b15322dc258a30/tools/code-quality.sh#L328-L330) script's exit logic:

```shell
if [ "${FAILED}" -ne 0 ]; then
    exit 1
fi
```

One of the earlier regression tests marked `FAILED=1` even though they reported "OK" in the output. Looking at the logs, the test that actually failed is:

**"Installer DNS environment failure regression"** (line 253 in code-quality.sh) - which shows:
```
FAIL: successful no-change recovery retained its snapshot
FAILED: Installer DNS environment failure regression
```

This test failure sets `FAILED=1`, causing the script to exit with code 1 at the end, even though subsequent checks (ShellCheck and shfmt) passed.

## Solution

You need to fix the **"Installer DNS environment failure regression"** test. The issue is that a snapshot from a successful no-change recovery operation isn't being retained properly. 

Check the test file at [tests/installer-dns-environment-failure.sh](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/e01455be6b2da14953feb31299b15322dc258a30/tests/installer-dns-environment-failure.sh) to verify:

1. **Snapshot retention logic**: Ensure that when a DNS environment setup completes successfully with no configuration changes, the snapshot state is properly saved and persisted.
2. **State recovery**: Verify that the test correctly validates that the snapshot from the no-change recovery scenario is retained across the test lifecycle.
3. **Cleanup procedures**: Check that temporary files and state markers are being cleaned up correctly without accidentally removing the snapshot that should be retained.

The formatting checks themselves are passing—this is a functional test failure in the DNS environment handling that needs to be debugged and fixed in the installer's snapshot management logic.
